src/main/scala/abalone/variant/Variant.scala

```diff
@@ -46,128 +46,233 @@ abstract class Variant private[variant] (
+        - push :
+          - dest contains a marble
+          - we have the direction :
+            - apply the direction from dest until there is an empty square or being off the board
+              - in case of being off the board, increment the score (capture = true)
+            - do the line move (orig to dest)
+        - side move :
+          - we received both directions :
+            - 1st direction : we know how to move through all these marbles
+            - 2nd direction : we know what direction we want to apply to all these 2 or 3 marbles
+  */
+  private def piecesAfterAction(pieces: PieceMap, orig: Pos, dest: Pos, directions: Directions): (PieceMap) = {
+    val lineDir: Option[Direction]    = directions.headOption
+    val sideDir: Option[Direction]    = directions.reverse.headOption
+    val origLineMove: Option[Pos]     = lineDir.flatMap(direction => direction(orig))
+    val origSideMove: Option[Pos]     = sideDir.flatMap(direction => direction(orig))
+    val destLineMove: Option[Pos]     = lineDir.flatMap(direction => direction(dest))
+    val destLineOneMove: Option[Pos]  = lineDir.flatMap(direction => if (destLineMove.isDefined) direction(destLineMove.get) else None)
+    val destLineTwoMove: Option[Pos]  = lineDir.flatMap(direction => if (destLineOneMove.isDefined) direction(destLineOneMove.get) else None)
+    val origDiagonalMove: Option[Pos] = (lineDir, origSideMove) match {
+      case ( Some(lineDir), Some(sideMove) ) => lineDir.apply(sideMove)
+      case _ => None
     }
-    if (player == P1)
-      (pieces, capture)
-    else
-      (pieces, capture)
+
+    if (directions.size == 0) // line move
+      pieces + (dest -> pieces(orig)) - (orig)
+    else if (directions.size == 1) // push
+      ( destLineMove, destLineOneMove, destLineTwoMove ) match {
+        case ( (None, _, _) ) => pieces + (dest -> pieces(orig)) - orig // __(_)o\
+        case ( (Some(destLinePos), _, _) ) if (!pieces.contains(destLinePos)) => pieces + (destLinePos -> pieces(dest)) + (dest -> pieces(orig)) - orig // __(_)o.
+        case ( (Some(destLinePos), None, _) ) => pieces + (destLinePos -> pieces(dest)) + (dest -> pieces(orig)) - orig // ___oo\
+        case ( (_, _, Some(desLineTwoPos)) ) => pieces + (desLineTwoPos -> pieces(dest)) + (dest -> pieces(orig)) - orig // ___oo.
+        case ( _ ) => pieces
+      }
+    else // side move
+      (origLineMove, origSideMove, origDiagonalMove)  match {
+        case ( Some(lineMovePos), Some(sideMovePos), Some(diagonalPos) ) =>
+          if (diagonalPos.index == dest.index) // oo
+            pieces + (sideMovePos -> pieces(orig)) - orig +
+              (diagonalPos -> pieces(lineMovePos)) - lineMovePos
+          else // ooo
+            pieces + (sideMovePos -> pieces(orig)) - orig +
+              (diagonalPos -> pieces(lineMovePos)) - lineMovePos +
+              (diagonalPos -> pieces(lineMovePos)) - diagonalPos
```

Comment:
is this correct? the first part of L256 and L257 are doing the same adding of `Pos`->`Piece`?
In general this whole function `piecesAfterAction` is complicated (it has to be) and i havent spent a lot of time to understand each piece of it. If you think it would be beneficial i suggest a call to review it together